### ISSUE TITLE
Don't create local SQL Database files inside the build workspace

### ISSUE THREAD

**Issue by MattGal** [dotnet/machinelearning#4175]
Code like this: https://github.com/dotnet/machinelearning/blob/master/test/Microsoft.ML.Tests/DatabaseLoaderTests.cs#L170

(where you're running tests as part of the build) has been leading to, when the build machine is reused, errors like:
```
Msg 5120, Level 16, State 101, Server <machine name>\LOCALDB#<somehex>, Line 1
Unable to open the physical file "F:\workspace.7\_work\1\s\bin\AnyCPU.Release-netcoreapp3_0\Microsoft.ML.Tests\netcoreapp3.0\TestDatabases\iris.mdf". Operating system error 3: "3(The system cannot find the path specified.)".
```
... this is because workspaces are automatically cleaned up (by deletion) but this leaves SQL server in a bad state.  Ideally this shouldn't happen at all, but if it must it'd be good to use the current user's profile directory;  I don't care where other than the workspace folder is not a good place for it.  This happens even to other teams' builds who also test using local SQL DBs 

### System information

- **OS version/distro**:  Windows Client and Server
- **.NET Version (eg., dotnet --info)**:  All (test bug)

### Issue

- **What did you do?** : Talk to other teams who need local SQL DBs to work
- **What happened?** : I found the code creating MDF files in a guaranteed-to-be-deleted place
- **What did you expect?** : Testing should occur outside builds.  Failing that, invariant cleanup should happen.  Failing that, put the files outside the workspace directory.

### Source code / logs

See https://github.com/dotnet/machinelearning/blob/master/test/Microsoft.ML.Tests/DatabaseLoaderTests.cs#L170.  When run as part of the build, the path this resolves is `bin\Arch.Config-TargetFramework\Microsoft.ML.Tests\netcoreapp3.0\TestDatabases\iris.mdf` inside the workspace.

**Comment by MattGal** [comment]
@tannergooding FYI, ping me if you have any questions.

**Comment by tannergooding** [comment]
@MattGal, could you help explain the problem?

The mdf file is placed in the test output directory (where the test executes from) by NuGet. It is then attached to using `AttachDbFilename` and the instance is cleaned up when the test completes.

I'm not seeing any problems cleaning this file up locally.

**Comment by MattGal** [comment]
Sure check out this build: https://dev.azure.com/dnceng/public/_build/results?buildId=332660 .  _Nobody_ has any difficulty it seems deleting that file, unfortunately, but once it is deleted without making the corresponding change in SQL Server it leaves things in an indeterminate state, and results in breaking other teams' builds.  Workspaces can and do get deleted after they are no longer needed in Azure DevOps, which is how we got to this state.

> ... and the instance is cleaned up when the test completes.

Is this true when the job times out or the XUnit process crashes?

**Comment by tannergooding** [comment]
> Is this true when the job times out or the XUnit process crashes?

.NET Core, AFAIK, doesn't provide any guaranteed cleanup state nor does a given CI job provide a guarantee that some cleanup job will definitely run. So I would guess the Dispose method and database disconnection doesn't happen if the app crashes.

You could have other potential, but possibly hidden, problems by moving where the file exists to workaround some file deletion error (file could be prevented from updated, be in some indeterminate state, etc).

There should likely be some pre-run cleanup that happens if it is needed (just like many of our repors kill build servers both before and after CI; the latter to be good citizens and the former to catch places where cleanup didn't happen previously)

**Comment by veikkoeeva** [comment]
A better solution would be not to use a database file explicitly. See at https://github.com/dotnet/machinelearning/issues/4156#issuecomment-527120149.

**Comment by tannergooding** [comment]
> but use the engine to create (and deletele) one dynamically

I'm not sure that will be scalable as more tests are getting added/created and larger datasets need to be tested.

**Comment by veikkoeeva** [comment]
@tannergooding How large are the datasets? You can set aside a backup or a snapshot and restore that when tests begin, perhaps. Or bulkoload the data in. It's close to that you do now, but managed by the database and probably faster too.

**Comment by ErikEJ** [comment]
@tannergooding EF Core team runs 1000ths of test that way.

**Comment by michaelgsharp** [comment]
Closing as I believe this was fixed with #5080
